xgboost_ray/main.py

```diff
@@ -359,28 +424,64 @@ def _handle_queue(queue: Queue, checkpoint: _Checkpoint,
             callback_returns[actor_rank].append(item)
 
 
-def _shutdown(remote_workers: List[ActorHandle],
+def _get_actor_alive_status(actors: List[ActorHandle],
+                            callback: Callable[[ActorHandle], None]):
+    """Loop through all actors. Invoke a callback on dead actors. """
+    obj_to_rank = {}
+
+    alive = 0
+    dead = 0
+
+    for rank in range(len(actors)):
+        actor = actors[rank]
+        if actor is None:
+            dead += 1
+            continue
+        obj = actor.pid.remote()
+        obj_to_rank[obj] = rank
+
+    not_ready = list(obj_to_rank.keys())
+    while not_ready:
+        ready, not_ready = ray.wait(not_ready, timeout=0)
+
+        for obj in ready:
+            try:
+                pid = ray.get(obj)
+                rank = obj_to_rank[obj]
+                logger.debug(
+                    f"Actor {actors[rank]} with PID {pid} is still alive.")
+                alive += 1
+            except Exception:
+                rank = obj_to_rank[obj]
+                logger.debug(f"Actor {actors[rank]} is _not_ alive.")
+                dead += 1
+                callback(actors[rank])
+    logger.info(f"Actor status: {alive} alive, {dead} dead "
+                f"({alive+dead} total)")
+
+    return alive, dead
+
+
+def _shutdown(actors: List[ActorHandle],
               queue: Optional[Queue] = None,
+              event: Optional[Event] = None,
               force: bool = False):
-    if force:
-        logger.debug(f"Killing {len(remote_workers)} workers.")
-        for worker in remote_workers:
-            ray.kill(worker)
-        if queue is not None:
-            logger.debug("Killing Queue.")
-            ray.kill(queue.actor)
-    else:
-        try:
-            logger.debug("Gracefully terminating workers")
-            [worker.__ray_terminate__.remote() for worker in remote_workers]
-            if queue is not None:
-                # Queues seem to stick around, so kill them always
-                logger.debug("Forcefully terminating queue")
-                ray.kill(queue.actor)
-        except RayActorError:
-            logger.warning("Failed to shutdown gracefully, forcing a "
-                           "shutdown.")
-            _shutdown(remote_workers, force=True)
+    for i in range(len(actors)):
+        actor = actors[i]
```

Comment:
for actor in actors?

Reply:
Since we're overwriting the array item here:
```
        actors[i] = None
``` we need the index anyway, so I'll probably keep it as is